File: 2307-replace-non-coprime-numbers-in-array/2307-replace-non-coprime-numbers-in-array.java
class Solution {
        public List<Integer> replaceNonCoprimes(int[] A) {
        LinkedList<Integer> res = new LinkedList();
        for (int a : A) {
            while (true) {
                int last = res.isEmpty() ? 1 : res.getLast();
                int x = gcd(last, a);
                if (x == 1) break; // co-prime
                a *= res.removeLast() / x;
            }
            res.add(a);
        }
        return res;
    }

    private int gcd(int a, int b) {
        return b > 0 ? gcd(b, a % b) : a;
    }
}

File: 2307-replace-non-coprime-numbers-in-array/2307-replace-non-coprime-numbers-in-array.cpp
class Solution {
public:
        vector<int> replaceNonCoprimes(vector<int>& A) {
        vector<int> res;
        for (int a: A) {   
            while (true) {   
                int x = gcd(res.empty() ? 1 : res.back(), a);
                if (x == 1) break; // co-prime
                a *= res.back() / x;
                res.pop_back();
            }
            res.push_back(a);
        }
        return res;
    }
};